src/EnergyPlus/DataHeatBalance.cc

```diff
@@ -2451,6 +2452,46 @@ namespace DataHeatBalance {
         }
     }
 
+    Real64 ConstructionData::setUserTemperatureLocationPerpendicular(Real64 userValue)
+    {
+        if (userValue < 0.0) {
+            ShowWarningError("Construction:InternalSource has a perpendicular temperature location parameter that is less than zero.");
+            ShowContinueError("Construction=" + this->Name + " has this error.  The parameter has been reset to 0.");
+            return 0.0;
+        } else if (userValue > 1.0) {
+            ShowWarningError("Construction:InternalSource has a perpendicular temperature location parameter that is greater than one.");
+            ShowContinueError("Construction=" + this->Name + " has this error.  The parameter has been reset to 1.");
+            return 1.0;
+        } else {    // Valid value between 0 and 1
+            return userValue;
+        }
+    }
+
+    void ConstructionData::setNodeSourceAndUserTemp(int &sourceNodeLocation,
+                                  int &userTempNodeLocation,
+                                  Array1D_int Nodes,
+                                  int NumOfPerpendNodes)
+    {
+        sourceNodeLocation = 0;
+        userTempNodeLocation = 0;
+        if (!this->SourceSinkPresent) return;
+
+        for (int Layer = 1; Layer <= this->SourceAfterLayer; ++Layer) {
+            sourceNodeLocation += Nodes(Layer);
+        }
+        
+        if ((sourceNodeLocation > 0) && (this->SolutionDimensions > 1)) sourceNodeLocation = ((sourceNodeLocation - 1) * NumOfPerpendNodes) + 1;
+
+        for (int Layer = 1; Layer <= this->TempAfterLayer; ++Layer) {
+            userTempNodeLocation += Nodes(Layer);
+        }
+        
+        if ((userTempNodeLocation > 0) && (this->SolutionDimensions > 1))
+            userTempNodeLocation = ((userTempNodeLocation - 1) * NumOfPerpendNodes)
+                                    + round(this->userTemperatureLocationPerpendicular * (NumOfPerpendNodes - 1)) + 1;
+            
+    }
```

Comment:
This is a good function, the only comment would be that some of the parentheses are unnecessary.

Reply:
@Myoldmopar Yeah, I admit it.  I tend to be really distinct with order of operations and overuse parentheses to make it crystal clear what I'm trying to do.  I can get rid of the extras.

Reply:
Changes made and they will show up in the next commit.